drivers/uc8151d/uc8151d.c

```diff
@@ -0,0 +1,320 @@
+    /* In the case the rising flank already hit, skip over the wait */
+    if ((bool)gpio_read(dev->params.busy_pin) == 0) {
+        /* Wait for the IRQ to unlock the mutex */
+        mutex_lock(&dev->gpio_wait);
+    }
+    /* Disable the IRQ again */
+    gpio_irq_disable(dev->params.busy_pin);
+}
+
+static void _gpio_cb(void *arg)
+{
+    uc8151d_t *dev = arg;
+
+    mutex_unlock(&dev->gpio_wait);
+}
+
+static void _cmd_start(uc8151d_t *dev, uint8_t cmd, bool cont)
+{
+    DEBUG("[uc8151d] cmd_start: command 0x%02x\n", cmd);
+    gpio_clear(dev->params.dc_pin);
+    spi_transfer_byte(dev->params.spi, dev->params.cs_pin, cont, (uint8_t)cmd);
+    gpio_set(dev->params.dc_pin);
+}
+
+static void _write_cmd(uc8151d_t *dev, uint8_t cmd,
+                       const uint8_t *params, size_t plen)
+{
+    spi_acquire(dev->params.spi, dev->params.cs_pin, SPI_MODE_0, dev->params.spi_clk);
+    _cmd_start(dev, cmd, plen > 0);
+    if (plen) {
+        spi_transfer_bytes(dev->params.spi, dev->params.cs_pin, false, params, NULL, plen);
+    }
+    spi_release(dev->params.spi);
+}
+
+static void _set_frame_rate(uc8151d_t *dev, uc8151d_frequency_t freq)
+{
+    uint8_t data = freq;
+
+    _write_cmd(dev, UC8151D_PLL_CTRL, &data, sizeof(data));
+}
+
+static void _reset_disp(const uc8151d_t *dev)
+{
+    for (unsigned i = 0; i < 3; i++) {
+        gpio_clear(dev->params.rst_pin);
+        ztimer_sleep(ZTIMER_MSEC, 10);
+        gpio_set(dev->params.rst_pin);
+        ztimer_sleep(ZTIMER_MSEC, 10);
+    }
+}
+
+static void _write_lut_data(uc8151d_t *dev, uint8_t command, const uc8151d_lut_phase_t *phases,
+                            size_t num)
+{
+
+    size_t padding_len = (36 - (sizeof(uc8151d_lut_phase_t) * num));
+
+    spi_acquire(dev->params.spi, dev->params.cs_pin, SPI_MODE_0,
+                dev->params.spi_clk);
+    _cmd_start(dev, command, true);
+    spi_transfer_bytes(dev->params.spi, dev->params.cs_pin,
+                       (bool)padding_len,
+                       phases,
+                       NULL,
+                       num * sizeof(uc8151d_lut_phase_t));
+
+    if (padding_len) {
+        for (size_t i = 0; i <  (padding_len - 1); i++) {
+            spi_transfer_byte(dev->params.spi, dev->params.cs_pin, true, 0x00);
+        }
+        spi_transfer_byte(dev->params.spi, dev->params.cs_pin, false, 0x00);
+    }
+    spi_release(dev->params.spi);
+}
+
+void _write_lut_partial_refresh(uc8151d_t *dev)
+{
+    _write_lut_data(dev, UC8151D_VCOM_LUT, &uc8151d_lut_partial_update_vcom, 1);
+    _write_lut_data(dev, UC8151D_W2W_LUT, &uc8151d_lut_partial_update_b2w, 1);
+    _write_lut_data(dev, UC8151D_B2W_LUT, &uc8151d_lut_partial_update_b2w, 1);
+    _write_lut_data(dev, UC8151D_W2B_LUT, &uc8151d_lut_partial_update_w2b, 1);
+    _write_lut_data(dev, UC8151D_B2B_LUT, &uc8151d_lut_partial_update_w2b, 1);
+}
+
+int uc8151d_init(uc8151d_t *dev, const uc8151d_params_t *params)
+{
+    memcpy(&dev->params, params, sizeof(uc8151d_params_t));
```

Comment:
Do we really need this memcpy? I would assume the params are static over the lifetime. (Damn, my brain starts thinking rust :crab:)